text/0000-reseeding-rand.md

```diff
@@ -0,0 +1,657 @@
+one go. Changing the `.fill_bytes` call to `.gen::<u32>()` makes `OsRng`
+250&times; slower than `ChaChaRng`.
+
+Relevant links for this RFC:
+
+- [C++11's `<random>` header][cpp_random]
+- [D's `std.random`][d_random]
+- ["Investigate replacing the ISAAC Rng with directly using the operating system,
+  or an eSTREAM Rng (etc.)"](https://github.com/rust-lang/rust/issues/10047)
+- ["Should RNGs really be clone?"](https://github.com/rust-lang/rust/issues/20973)
+
+[cpp_random]: http://en.cppreference.com/w/cpp/numeric/random
+[d_random]: http://dlang.org/phobos/std_random.html
+
+## Cryptographic security
+
+The only interface that will be recommended as cryptographically
+secure will be the RNG that interfaces with the operating system. This
+will be crystal clear in the documentation.
+
+# Detailed design
+
+The stable-at-1.0 parts of the module will essentially just consist of
+2 traits (plus one extension trait) and 3 RNG algorithms (plus one
+wrapper).
+
+## Traits
+
+`std::rand` will include the following generic set-up;
+
+```rust
+/// Things that can be created randomly, where the creation can be
+/// mediated by information contained in `Data`.
+trait Random<Data = FullRange> {
+    /// Create a random value of type `Self`
+    fn random<R: ?Sized + RngBase>(data: Data, rng: &mut Rng) -> Self;
+}
+
+// Example:
+
+// create a raw u8 with no restriction.
+impl Random<FullRange> for u8 {
+    ...
+}
+
+// create a u8 between the bounds of range (same exclusivity properties as usual)
+impl Random<Range<u8>> for u8 {
+    ...
+}
+
+
+/// Random number generator.
+trait Rng: io::Reader {
+    /// Generate/fetch the next `u32` (no range restriction)
+    #[unstable]
+    fn next_u32(&mut self) -> u32 { ... }
+
+    /// Generate/fetch the next `u64` (no range restriction)
+    #[unstable]
+    fn next_u64(&mut self) -> u64 { ... }
+    /// Generate/fetch the next `f32` in the range [0,1)
+    #[unstable]
+    fn next_f32(&mut self) -> f32 { ... }
+    /// Generate/fetch the next `f64` in the range [0,1)
+    #[unstable]
+    fn next_f64(&mut self) -> f64 { ... }
+}
+
+impl<'a, T: ?Sized + Rng> Rng for &'a mut T {}
+impl<T: ?Sized + Rng> Rng for Box<T> {}
+
+/// Basic extension functionality to ensure that `Rng` is object-safe.
+trait RngExt {
+    fn gen<T, D>(&mut self, data: D) -> T where T: Random<D> { ... }
+
+    /// Return an iterator that generates an infinite sequence of random values of type `T`.
+    fn gen_iter<T, D>(self, data: D) -> Generator<T, D, Self> where T: Random<D> { ... }
+
+    /// Randomize the order of the elements of `values`
+    fn shuffle<T>(&mut self, values: &mut [T]) { ... }
+    /// Select one value uniformly at random from the slice.
+    #[unstable]
+    fn choose<'a>(&mut self, values: &[T]) -> Option<&'a T> { ... }
+    /// Select `n` values uniformly at random from the iterator.
+    fn sample<I: Iterator>(&mut self, n: uint, values: I) -> Vec<I::Item> { ... }
```

Comment:
The implementation has to allocate internally regardless, since it has to buffer until it exhausts the iterator.